#28: Cleaning up a little to enable unit-testing API code.

kocielnik opened the pull request:
If you feel uncomfortable with either of these changes, please let me know! I can revert such a change or give more context why I think it may make sense.

Included propositions:

1. Replace CircleCI config with GitHub Actions config.
    - I've tried to mimic the CircleCI config one-to-one with GitHub Actions.
    - Motivation: I did not have a CircleCI account when I forked this repo. That required me to create an account with CircleCI. To use GitHub Actions, I only need the GitHub account.
        - Counter-arguments:
            1. CircleCI advertises itself as "the fastest CI service". It may be GitHub Actions are slower. Comments are welcome.
            2. Some people may *not* have a GitHub account, while having a CircleCI account. For them, this change would actually make things *more* complicated.

2. Move requirement specification to `pyproject.toml`.
    - This allows using Poetry to enter a virtual env with all deps installed using one command.
    - PIP is still able to install this package from `pyproject.toml` - this support was added a while ago and is now pretty standard.
        - GitHub Actions CI config is an example of how similar the installation process is, while allowing the developer more flexibility (`poetry install` to install all deps, `poetry shell` to enter the virtualenv shell with them installed).
    - Counter-argument: For someone not familiar with Poetry, other workflows and tools may be much more tempting. The `pyproject.toml` file is pretty standard now as the format for describing packages, but the sections to be used depend on one's favourite build tool.

3. Move code from `main.py` to `src/server.py` for testing.
    - This required moving also templates to `src/templates`.
        - Pro: root repo dir looks cleaner.
        - Con: I may not have full understanding of consequences of moving these templates to a subdirectory. I confirmed the page worked fine when hosted locally.

4. Untrack `.envrc`.
    - This file tells your shell to enter the virtualenv immediately after `cd`-ing to the package directory.
    - The tool that relies on this file is called Direnv.
    - It loads the virtualenv in two seconds, unloads in less than a second.

kocielnik committed:
Add an API test.

kocielnik committed:
Try to enable GitHub Actions.

kocielnik committed:
Try to remove the Werkzeug warning.

kocielnik committed:
Get a real quote.

kocielnik committed:
Rename package to src for now.

While pyproject.toml must contain the package name, using a name other
than "src" would require cascading changes to the submodules.

Let's leave that for a later stage, if required.

kocielnik committed:
Allow Python 3.10.

kocielnik committed:
Allow Python ^3.8.

Recent Flask requires ^3.8.

kocielnik committed:
Skip checking for requirements.txt (extra step).

kocielnik committed:
Remove requirements.txt.

kocielnik committed:
Add PyTest as a plain dependency.

kocielnik committed:
Specify Mark as the author.

kocielnik committed:
Refactor: Move server logic from main.py to src/server.py.

kocielnik committed:
Debug: Print current values.

kocielnik committed:
Remove a debug print.

kocielnik committed:
Remove CircleCI config.

Configured GitHub Actions instead.

If that isn't equivalent, we can revert.

kocielnik committed:
Untrack .envrc.

kocielnik committed:
Remove .envrc.

mrhappyasthma commented:
Thanks for all the work on this! I'll try to review it this weekend when we have some time, but generally I'm okay with any improvements to the project :)

1 - I don't have a strong preference. If we can do it as part of github actions, that should be fine. I just did Circle CI since I was familiar with it from other work.

2 - I'm not a python programmer by trade, so I never even heard of `pyproject` haha. But it seems to be the standard per this - https://www.reddit.com/r/learnpython/comments/1bmxe6i/whats_the_difference_between_pyprojecttoml/. So let's do it.

3 - Seems fine. I don't have a strong opinion and `src/server` is better named than `main`.

4 - Seems reasonable to me.

kocielnik commented:
Thanks for precise feedback right on the following day!

Regarding point 1, imagining we have GitHub Actions configured, we could next try out a Google-backed recipe to deploy to Google App Engine straight from GitHub using Actions at https://github.com/google-github-actions/deploy-appengine.

kocielnik committed:
Add PyLint exclusions for most common warnings.

These specific warnings were so numerous they obscured the code.

Before anyone gets to cleaning these warnings [1], let's disable them
for now just to make reading the code easier.

[1]: *If* they should ever be cleaned at all.

mrhappyasthma commented on test_api.py:
```diff
@@ -0,0 +1,27 @@
+
+
+def mock_fetch_data():
+    return {}
```
Is this used?

kocielnik replied:
Nope, good note. Removed!

mrhappyasthma commented on test_api.py:
Should this file go in the tests/ dir?

kocielnik replied:
Yes, that will be much better. Moved!

mrhappyasthma commented on test_api.py:
```diff
@@ -0,0 +1,27 @@
+def test_import_app():
+    app = create_app(fetchDataForTickerSymbol)
+
+    with app.test_client() as test_client:
```
I'm not too familiar with writing tests for Flask (i'm not a python dev by trade, heh).

This is for hermetic tests, where we provide mock responses. Right?

(I also imagine a world where we write some tests that actually fetch data, which may be flaky as they depend on network. but might work as a good way to detect 'is this endpoint still live') I think ideally we want both:

1. Does our parsing work given canned data (hermetic, quick, reliable)

2. Does fetching the data over the network still reply with a payload that has valid data and not 404 or a different data format.

kocielnik replied:
> This is for hermetic tests, where we provide mock responses. Right?

Yes.

> I think ideally we want both

I agree. I may even try to include these two kinds of tests here in this PR. I have an idea.

mrhappyasthma replied:
I think it's alright to do it in a separate PR just to keep the scope small.

This one looks good so let's land it.

kocielnik committed:
Move test_api to tests/.

kocielnik committed:
Remove unneeded mock.

ThePatricide commented:
@kocielnik my apologies as I'm fairly new to git, python and development, but did this PR replace the requirements.txt with pyproject.toml? The README now seems outdated and I'm unsure how to install the dependencies now. Could you perhaps point me in the right direction?

Edit: I think I figured it out. It is just replacing the last part of -r requirements.txt by . (a dot).

bioluminesceme commented:
@ThePatricide 

    pip install flask lxml requests-futures

kocielnik commented:
@benninkcorien, @ThePatricide – thanks for sharing your insights!

I've created a PR, referenced above, to update the README.
